src/main/scala/co/teapot/graph/MemoryMappedDirectedGraph.scala

```diff
@@ -67,350 +93,117 @@ class MemoryMappedDirectedGraph(file: File) extends DirectedGraph {
-    val nodeCountInChunk = math.min(nodesPerChunk, nodeCount - chunkIndex * nodesPerChunk)
-    val neighborArrays = Array.fill(nodeCountInChunk)(new IntArrayList())
-    // The difference between a nodeId and the corresponding index into neighborArrays
-    val nodeOffset = chunkIndex * nodesPerChunk
-    val tempFileEdgeCount = tempFile.length() / 8L
-    for (edgeIndex <- Stream.range(0L, tempFileEdgeCount)) {
-      val id1 = inStream.readInt()
-      val id2 = inStream.readInt()
-      neighborType match {
-        case GraphDir.Out => neighborArrays(id1 - nodeOffset).add(id2)
-        case GraphDir.In => neighborArrays(id2 - nodeOffset).add(id1)
+  /** Loads segments into memory-mapped buffers. */
+  private def memoryMapSegments(dir: EdgeDir): (Array[Segment]) = {
+    (0 until segmentCount).toArray map { i =>
+      val baseOffset = dir match {
+        case EdgeDir.Out => OffsetToSegmentOffsets
+        case EdgeDir.In => OffsetToSegmentOffsets + (segmentCount + 1) * BytesPerSegmentOffset
       }
+      val offset = headerData.getLong(baseOffset + i * BytesPerSegmentOffset)
+      val size = headerData.getLong(baseOffset + (i + 1) * BytesPerSegmentOffset) - offset
+      val buffer = fileChannel.map(MapMode.READ_ONLY,  offset, size)
+      new Segment(buffer, nodesPerSegment)
     }
-    inStream.close()
-    neighborArrays
   }
+}
 
-  /** Helper method to edgeFileToGraph.
-    * Appends neighbor data to binaryGraphOutput, and writes the offsets to binaryGraphChannel at
-    * chunkOffsetsStart.
-    */
-  def writeNeighborDataToGraphAndReturnOffset(
-      neighborLists: IndexedSeq[IntArrayList],
-      initialCumulativeNeighborOffset: Long,
-      binaryGraphOutput: DataOutputStream,
-      binaryGraphChannel: FileChannel,
-      chunkOffsetsStart: Long,
-      isLastChunk: Boolean): Long = {
-    var cumulativeNeighborOffset = initialCumulativeNeighborOffset
-    val neighborOffsets = new Array[Long](neighborLists.size)
-    for ((neighbors, i) <- neighborLists.zipWithIndex) {
-      neighborOffsets(i) = cumulativeNeighborOffset
-      val sortedDistinctNeighbors = IntArrayUtil.sortedDistinctInts(neighbors)
-      for (j <- 0 until sortedDistinctNeighbors.size()) {
-        val neighborId = sortedDistinctNeighbors.get(j)
-        // binaryGraphOuput is buffered, so the writeInt calls are efficient
-        binaryGraphOutput.writeInt(neighborId)
-        cumulativeNeighborOffset += 4L
-      }
-    }
-    // Store the ending offset for the last node
-    val finalOffset = cumulativeNeighborOffset
-    binaryGraphOutput.flush()
-
+/* Reads data for a segment of nodes. Methods are in terms of node index i = id /
+segmentCount, where id is the id of the node in the overall graph.  The same segment class is
+used for out-neighbors and in-neighbors.
+ */
+private[graph] class Segment(val data: MappedByteBuffer, nodesPerSegment: Int) {
+  val neighborDataOffset = (nodesPerSegment + 1) * BytesPerNodeOffset
```

Comment:
Sure.  That makes the addressing cleaner (in C++ language, this lets us do pointer arithmetic on int* rather than char*).  In that case I'm inclined to just drop IntPerNodeOffset and assume nodeOffsets are Ints.  (ByteBufferIntSlice only returns Ints, so I don't think IntPerNodeOffset != 1 would be possible anyway).  
I don't understand how this helps with OutOfBoundsExceptions--I think the ByteBuffer I'm using will already check if I try to get an Int past the end of the segment.